alvr/server_core/src/tracking/mod.rs

```diff
@@ -65,48 +74,128 @@ impl TrackingManager {
             last_head_pose: Pose::IDENTITY,
             inverse_recentering_origin: Pose::IDENTITY,
             device_motions_history: HashMap::new(),
+            recentering_marker: None,
+            markers: HashMap::new(),
             hand_skeletons_history: [VecDeque::new(), VecDeque::new()],
             max_history_size,
         }
     }
 
-    pub fn recenter(
-        &mut self,
-        position_recentering_mode: PositionRecenteringMode,
-        rotation_recentering_mode: RotationRecenteringMode,
-    ) {
-        let position = match position_recentering_mode {
-            PositionRecenteringMode::Disabled => Vec3::ZERO,
-            PositionRecenteringMode::LocalFloor => {
-                let mut pos = self.last_head_pose.position;
-                pos.y = 0.0;
-
-                pos
+    pub fn recenter(&mut self, mode: &RecenteringMode) {
+        let recentering_origin = match mode {
+            RecenteringMode::Stage {
+                marker_based_colocation,
+            } => {
+                if let Some(colocation_config) = marker_based_colocation.as_option() {
+                    if let Some(marker_pose) = self.markers.get(&colocation_config.qr_code_string) {
```

Comment:
This should really be
```rs
let Some(colocation_config) = ... else {
   // ...
   self.inverse_recentering_origin = Quat::IDENTITY;
    return;
};

let Some(marker_pose) = ... else {
   // ...
   return;
}
```
Since that greatly reduces indentation and the complexity of tracking what the fallback is.
And I think manually setting `inverse_recentering_origin` isn't a big issue in terms of someone modifying the epilogue of the function in the future (the `return;` already skips stuff that could happen in the epilogue as-is).

But if you do care I'd suggest extracting the marker logic into a second function (that might be a good idea in general) that returns an option and doing the following here:
```rs
if let Some(origin) = self.calculate_marker_origin() {
   origin
} else {
   return;
}
```